File: March 2025/29.cpp
class Solution {
  public:
    vector<int> jobSequencing(vector<int> &deadline, vector<int> &profit) {
        int n = deadline.size(); // Number of jobs
        vector<pair<int, int>> jobs;
        
        // Combine deadlines and profits into a vector of pairs
        for (int i = 0; i < n; i++) {
            jobs.push_back({profit[i], deadline[i]});
        }
        
        // Sort jobs in decreasing order of profit
        sort(jobs.rbegin(), jobs.rend());
        
        // Find the maximum deadline
        int maxDeadline = 0;
        for (int d : deadline) {
            maxDeadline = max(maxDeadline, d);
        }
        
        // Parent array for union-find
        vector<int> parent(maxDeadline + 1);
        for (int i = 0; i <= maxDeadline; i++) {
            parent[i] = i;
        }
        
        // Function to find the parent of a slot (using path compression)
        function<int(int)> findParent = [&](int x) -> int {
            if (parent[x] == x) return x;
            return parent[x] = findParent(parent[x]);
        };
        
        int maxProfit = 0, jobsDone = 0;

        // Iterate through the jobs
        for (auto &job : jobs) {
            int p = job.first;
            int d = job.second;

            // Find the latest available slot
            int availableSlot = findParent(d);
            if (availableSlot > 0) {
                parent[availableSlot] = findParent(availableSlot - 1); // Mark the slot as taken
                maxProfit += p;
                jobsDone++;
            }
        }

        return {jobsDone, maxProfit};
    }
};

File: March 2025/29.py
class Solution:
    def jobSequencing(self, deadline, profit):
        # Combine jobs with their deadlines and profits
        jobs = list(zip(deadline, profit))
        
        # Sort jobs in decreasing order of profit
        jobs.sort(key=lambda x: x[1], reverse=True)
        
        # Find the maximum deadline
        max_deadline = max(deadline)
        
        # Create an array to track free time slots
        parent = list(range(max_deadline + 1))
        
        def findParent(x):
            # A function to find the parent slot using path compression
            if parent[x] == x:
                return x
            parent[x] = findParent(parent[x])
            return parent[x]

        max_profit = 0
        jobs_done = 0

        # Iterate through the sorted jobs
        for d, p in jobs:
            # Find the latest available slot before or at the deadline
            available_slot = findParent(min(d, max_deadline))
            if available_slot > 0:  # A slot is available
                parent[available_slot] = findParent(available_slot - 1)  # Mark the slot as taken
                max_profit += p
                jobs_done += 1

        return [jobs_done, max_profit]
